### ISSUE TITLE
Package for macOS

### ISSUE THREAD

**Issue by cecilios** [lenmus/lenmus#103]
Hi Bob, 
I have opened this issue for better management. I receive many emails and it is difficult to find, after some time, the email in which something was discussed. So, please let's continue talking here. Thank you.

> The instructions are pretty clear.  I've cloned the source from GitHub, installed CMake and wxWidgets and am looking at some of the other dependencies. Some, such as Zlib and FreeType2 are already included with Mac OS X.

Glad to know! The wxWidgets is always the hardest part with the dependencies. 

> I'm keeping notes of what I'm doing in case you want to include compilation instructions for Mac on your site.

Yes, please. Not only for publication but also for reference for other possible contributors using Mac computers. Thank you.

Once you have installed all dependencies, I would suggest to try to build using the command line and cmake, to see what errors come out. Open a terminal and type:
```
     cd <your-projects-folder>/lenmus
     mkdir z_mybuild
     cd z_mybuild
     cmake ../
```
Probably no errors will come out. Go telling me what's going on and I'll try to help you how to solve the issues.

Once cmake generates without errors, you can move on to compile and build:
```
     make -j4      #change the '4' for the num. of processors of your machine
```
Here, I'm sure errors will appear.

For fixing the errors and debugging it is better to use XCode. The most "complex" task to finish setting the development environment would be to create a project file for XCode, so that you can compile and debug from the IDE. I don't know if XCode can create a project file by importing the cmake file. In any case, it would be better to create the project file by starting a new project, including the LenMus sources and configuring the dependencies.

Once you are able to compile from XCode, the hardest part should be done.

**Comment by blurbler27** [comment]
Hi Cecilio

Sorry this is taking a little longer than I thought.

I have the dependencies installed, and working.  That is, I can successfully compile programs using the dependencies.

I need a little guidance with CMake.   The first error from cmake that I didn't know how to fix concerned finding a library.  Cmake terminated telling me that it couldn't find ZLib.  Zlib is included with Apple Developer tools,  and a program including zlib.h will correctly compile with "gcc test.c -lz".

If Cmake can't find this library, I'm sure it will have trouble with others, so if I can solve this I'll probably make a lot of progress.  Any hints on configuring cmake?  (Documentation, etc.)

Thanks,
Bob

**Comment by cecilios** [comment]
Hi Bob,
> this is taking a little longer than I thought.

Don't worry. It is always like this and we are in no rush!

> I have the dependencies installed, and working.

Great! Good work!

> The first error from cmake that I didn't know how to fix concerned finding a library

That's the most common problem with cmake because, contrary to what cmake seems to advertise, macro `find_package` finds nothing unless you tell cmake were to look for it.

For instance, referring to file CMakeLists.txt, when anyone reads these lines
https://github.com/lenmus/lenmus/blob/c8afc761d4a99b257725e3bfcdf9081ef5feba9f/CMakeLists.txt#L684-L687

the normal expectation is to think that cmake has methods for looking for libraries. Wrong! 

`find_package(XXXXX)` just invokes a module named `FindZlib.cmake`. This module must be provided either by cmake or by your application. And normally, the module provided by cmake, if exists, doesn't work and needs some changes. So, in practice, you are forced to write your own `FindXXXXX.cmake` modules. For lenmus, they are in folder `cmake-modules`.

Now look at the module `FindZlib.cmake`:
https://github.com/lenmus/lenmus/blob/master/cmake-modules/FindZLib.cmake
It mainly just contains a list of paths were, normally, it can be expected to find zlib installed.

So, for fixing this, it is just enough to determine the standard places were zlib can be installed in macOS and to add these paths to the list. 

To my taste, official documentation for cmake ( https://cmake.org/documentation/ ) is very bad. It is useful only to know the available functions and their parameters. For nothing else. If you have problems with FindZlib it is usually better to google for FindZlib. You will find a lot of modules from other projects and, with luck, you will find the paths they include for macOS.

Hope this helps!

**Comment by cecilios** [comment]
An important thing: you need to install not only the zlib library but also the headers. I don't know in macOS but in linux its common to have two different installation packages. For instance:
`zlib1g-dev.deb` for the library plus it headers and `zlib1g.deb` for only the library.

If you have the library installed but not the headers, cmake will always report that the library is not found.

**Comment by blurbler27** [comment]
I've made a lot of progress.  I'm keeping notes and will send you a complete report.

So far I've only had to change one file in the LenMus source (in cmake-modules/FindZlib.cmake) and I don't think this change will hurt compilation on any other platform.  Running "make" generates a lot of warnings which I'm saving in log files for later investigation, but gets a long way before terminating with an error.

I compiled wxWidgets myself, because the homebrew and MacPorts packages are for an older version, and I assumed the most recent version was preferred.  However, when trying to compile LenMus, I'm getting an error
        /Users/rbyerly/devel/LenMus/lenmus/src/updater/lenmus_updater.cpp:257:5: error: 
       unknown 
           type name 'wxDialUpManager'
              wxDialUpManager* pManager = wxDialUpManager::Create();

In the default setup.h for the OS X version of wxWidgets, the preprocessor variable wxUSE_DIALUP_MANAGE is set to 0, which suppresses compilation of the wxDialUpManager class. When I tried to Google the issue, I found one old post (https://forums.wxwidgets.org/viewtopic.php?t=28734) which says that the DialUpManager is not supported on OS X platforms.  This post is 10 years old.  I have not yet tried the effect of changing the setup.h file for OS X in wxWidgets to set wxUSE_DIALUP_MANAGE to 1.  I guess I could try this.  But if it really isn't supported this will be futile.

Before I try this, I wanted to ask how important the wxDialUpManager is  to LenMus. If most users have a stable internet connection, probably not very important.  But I don't want to assume this on my own.  It looks like it would be easy to edit lenmus_updater.cpp to avoid the use of wxDialUpManager.  I'm going to try this (just have the Updater::CheckInternetConnection method always return true) just to see if everything else works OK on Macs.  If everything else works OK we can think about the best way to fix this issue.

Regards,
Bob

**Comment by cecilios** [comment]
> I've made a lot of progress.

Glad to know! Good job.

> Running "make" generates a lot of warnings which I'm saving in log files for later investigation

Don't worry about them for know. Probably stupid things such as conversion from one type to another. we will fix them later.

> I compiled wxWidgets myself, because the homebrew and MacPorts packages are for an older version, and I assumed the most recent version was preferred.

I'm currently using 3.0.4 and most distros has that one or 3.0.5. But a more recent version is good for testing. For preparing a macOS package this is not importtant as every dependency will be included in the package. Better if we have instructions from building wxWidgets from sources. Less problems to automate building the macOS package.

> I have not yet tried the effect of changing the setup.h file for OS X in wxWidgets to set wxUSE_DIALUP_MANAGE to 1. I guess I could try this. But if it really isn't supported this will be futile. 

I agree, it is better to replace it. Do not spend time on trying to build wxWidgets with support for wxDialUpManager.

> I wanted to ask how important the wxDialUpManager is to LenMus.

It is only used to check from time to time if a new version of LenMus is available. It is not important. We will re-code the updater when everything else works OK and a package can be generated.

**Comment by blurbler27** [comment]
Everything compiles and I can run the executable.  I'm not getting a nice opening screen as in the Linux version, though, and when I try to open a book I get an error message:
          Line 0. Invalid real number '20.0f'. Replaced by '0'.

I think I'll first have a look at the warning messages from "make".  I saved make output using "make 2>&1 > make.log" so I can examine these more carefully.  Or when I improve my Xcode knowledge, maybe I can use the Mac's debugging tools.  (I'm still fairly new to Xcode. I've mostly just used the command line tools.)

Judging from the Linux version, this looks like a very nice program. I hope it can be ported successfully to OS X.

**Comment by cecilios** [comment]
> Everything compiles and I can run the executable

Excellent! Great job!

> I'm not getting a nice opening screen as in the Linux version

I'm curious, could you post a screenshot?

> I try to open a book I get an error message: Line 0. Invalid real number '20.0f'. Replaced by '0'.

It is a message from the eBooks parser, probably from this method:
https://github.com/lenmus/lenmus/blob/c8afc761d4a99b257725e3bfcdf9081ef5feba9f/lomse/trunk/src/parser/lmd/lomse_lmd_analyser.cpp#L356-L368

It looks as if the value `20.0f` is bad in an eBook. It should be `20.0`. Which book were you opening?

After the error, does the program displays anything? could you post a screenshot?

> I think I'll first have a look at the warning messages from "make". I saved make output using "make 2>&1 > make.log"

Please, send me a copy. I will also look at this.

Thank you!

**Comment by blurbler27** [comment]
When I start LenMus nothing appears on the mains screen, although a menu appears in the command bar.
![Fullscreen_4_29_20__11_42_AM](https://user-images.githubusercontent.com/32274358/80623020-3b2a6c80-8a0f-11ea-8320-26b105f4c3e7.png)

If I try to open the books:
![Fullscreen_4_29_20__11_50_AM](https://user-images.githubusercontent.com/32274358/80623478-d6234680-8a0f-11ea-9b78-1689e62e3921.png)

When I try to open Theory and Harmony I get this error:
![Fullscreen_4_29_20__11_44_AM](https://user-images.githubusercontent.com/32274358/80623530-efc48e00-8a0f-11ea-86d9-94771ee470cc.png)

When I close this window I just get a big blank window.  I can terminate lenmus from the menu bar.

I'm including a "lomse-log.txt" from the last run, my "make.log", and "install.log".

You mentioned that you were using version 3.0 of wxWidgets.   I'm using 3.1.  I can try using the 3.0 version from homebrew if you think that would help.

[lomse-log.txt](https://github.com/lenmus/lenmus/files/4553116/lomse-log.txt)
[make.log](https://github.com/lenmus/lenmus/files/4553127/make.log)
[install.log](https://github.com/lenmus/lenmus/files/4553129/install.log)

**Comment by cecilios** [comment]
Thank you for the images. That helps me to understand what is going on.

> You mentioned that you were using version 3.0 of wxWidgets. I'm using 3.1. I can try using the 3.0 version from homebrew if you think that would help.

No please, continue using 3.1. I think the black window has nothing to do with this, but with the program not being fully prepared for macOS. We need to review this:
https://wiki.wxwidgets.org/WxMac-specific_topics

I'm going to study the logs and the above link and I will tell you something tomorrow.

**Comment by blurbler27** [comment]
Great.  The wxwidgets link you gave looks interesting.  I'll have a closer look at it as well.

**Comment by cecilios** [comment]
I've seen the logs and apart from all the build warnings I've found nothing important.

I've checked and I can confirm that the _“Line 0. Invalid real number '20.0f'. Replaced by '0'.”_ error is caused by a bad number in the eBook sources. I will fix this but it is not important and, apart from displaying the error message, the program will work OK.

So, the most urgent task would be to solve the blank screen issue. I have no idea of what is the problem, so I've thought two possibilities:

1. An error in lenmus code. After checking the lenmus code and the wxWidgets samples, the only possibility I can think is that the blank screen could be caused by lines:
https://github.com/lenmus/lenmus/blob/c8afc761d4a99b257725e3bfcdf9081ef5feba9f/src/app/lenmus_app.cpp#L976-L981

by not doing `m_frame->Show(true);`. So, test 1: In file src/app/lenmus_app.cpp, comment out line 977 (`#ifndef __WXMAC__`) and line 979 (`#endif //ndef __WXMAC__`)

2. The other possibility is to follow this suggestion in:
https://wiki.wxwidgets.org/WxMac-specific_topics#My_app_can.27t_be_brought_to_the_front.21

So, test 2: In file src/app/lenmus_app.cpp
https://github.com/lenmus/lenmus/blob/c8afc761d4a99b257725e3bfcdf9081ef5feba9f/src/app/lenmus_app.cpp#L45-L49
add after line 46 the following lines:
```
#ifdef __WXMAC__
    #include <ApplicationServices/ApplicationServices.h>
#endif
```

and after line 147
https://github.com/lenmus/lenmus/blob/c8afc761d4a99b257725e3bfcdf9081ef5feba9f/src/app/lenmus_app.cpp#L146-L151
add the following code:
```
#ifdef __WXMAC__
    ProcessSerialNumber PSN;
    GetCurrentProcess(&PSN);
    TransformProcessType(&PSN,kProcessTransformToForegroundApplication);
#endif
```


---------------
The third test I suggest would be to follow the solution proposed here:
https://wiki.wxwidgets.org/WxMac-specific_topics#Building_a_MacOSX_application_bundle

For this it would be necessary to modify the file CMakeLists.txt. I will try do it.

**Comment by blurbler27** [comment]
I'll try these suggestions.  For the third test, I will remark that when I tried the sample/demo programs included with wxwidgets, they seemed to run fine when invoked from the command line.  Building an application bundle is a good goal, though.

**Comment by cecilios** [comment]
I've looking at how the samples are build and they create a boundle, as explained here:
https://wiki.wxwidgets.org/WxMac-specific_topics#Building_a_MacOSX_application_bundle

But to do it we need to modify the file CMakeLists.txt so that cmake do it. I'm now investigating how to modify that file.

**Comment by blurbler27** [comment]
Gah!!!  Ignore last message.  I tried running LenMus while still in the build directory.  When I tried it again from my home directory I got a more normal looking screen.  I'll have to put this aside for now, but I'll play with it when I have time and figure out what functionality is working and what isn't.

Bob
![NOS](https://user-images.githubusercontent.com/32274358/80718199-75534700-8abf-11ea-805b-31c922ab1f32.png)

**Comment by cecilios** [comment]
Great! This is advancing at good pace.

I've been investigating about the OSX application bundle. It is a tree of nested folders:
- The root folder is <yourApp>.app
- The executable is placed in “Contents/MacOS/<executable>”
- The application resources can be placed as separate files in "Contents/Resources"
- Additional metadata can be contained in "Contents/Info.plist" and "Contents/version.plist"
- Locale/Language-specific data is contained in "Contents/Resources/<Language>.lproj" when managed by macOS, but can be in other places when managed by your application
- Dependent dynamic libraries and frameworks (e.g. wxWidgets) can also be placed in the bundle folders

For lenmus it should be something as:
```
/Applications
    └── lenmus.app/
        └── Contents/
            ├── Info.plist      (generated by cmake. A template included)
            ├── version.plist   (should be generated by cmake)
            ├── PkgInfo (it only contains "APP???"  (echo "APPL????" > lenmus.app/Contents/PkgInfo)
            ├── Frameworks/
            │   └── <libname>.dylib (e.g.: wxWidges libraries)
            ├── MacOS/
            │   └── lenmus (the executable)
            └── Resources/      (copied here by “make install”)
                ├── locale/
                ├── res/
                ├── templates/
                ├── xrc/
                ├── AUTHORS
                ├── CHANGELOG.md
                ├── .. etc.
                │
```

So, the changes to do in CMakeLists.txt are simple, just to change some install folders and add a couple of new files (e.g. Info.plist)

Before changing CMakeLists.txt I would like to update it with the changes you had to do for Zlib. Could you please upload your current file to your remote repo? I will update mine from there. If you need help for doing that, just open a terminal window and do:
```
# move to lenmus local repo
cd /Users/rbyerly/devel/LenMus/lenmus

# display the files that you added or changed
# CMakeLists.txt should be in the list
git status 
    #the output should be something as:    
    Changes not staged for commit:
      (use "git add <file>..." to update what will be committed)
      (use "git checkout -- <file>..." to discard changes in working directory)

	    modified:   CMakeLists.txt
        ...

    Untracked files:
      (use "git add <file>..." to include in what will be committed)

        ...


# add the file to the list of files to commit
git add CMakeLists.txt

# commit your changes
git commit -m"Changes to build in macOS"

# upload your changes to your remote repo at GitHub
git push
```

Other things to investigate:
- Why the background is black instead of white? Is it black when you run the wxWidgets samples?
- Test the sound: Run lenmus, go to lenmus main menu and chose _Sound > Test sound_. Do you hear anything?

**Comment by blurbler27** [comment]
OK. I'm uploading the files.  

There are two files I had to change in such a way they would not be suitable for other platforms.  I'm uploading them anyway so that you can take a look:
1.  CMakeLists.txt:  I commented out lines 1266 to 1280, since they had to do with installing icons, Desktop entries, and AppStream upstream metadata.  If these are relevant to OS X, they don't go in /usr/share, especially since Apple doesn't allow users or non-apple packages to install there.  This is in a "if(UNIX)..." block.  Perhaps the "if" statement could be changed to something like "if(UNIX and not APPLE)..."?  I don't know the correct syntax.
2.  In "src/auxmusic/lenmus_cadence.cpp", for some reason that I totally do not understand, the Mac C preprocessor was not recognizing the syntax of the line "#if (LENMUS_COMPILER_MSVC == 1)     //VC++ 2003 does not support variadic macros".  Since this was not relevant to OS X, I just edited this file to avoid this macro.  I'll continue to play with this to see if I can find a variant that Mac likes but which presumably will also work on other platforms.

Other questions:
A.  The background was black instead of white because I had my Mac set in Dark Mode.  When I change it to Light Mode the  background is white.
B.  Sound works.  I haven't given the program a thorough workout yet, but I will.

I have some difficulty resizing and iconizing windows.  This must be a wxWidgets thing.  I'll see if I can find more information on this.

**Comment by blurbler27** [comment]
I forgot to mention:  I still get that annoying error window when I try to open a book, but when I click the OK button the book shows up anyway.  And I can do the exercises if I work around this.

**Comment by cecilios** [comment]
This is going faster than I expected! We are starting to see the light at the end of the tunnel!

>  The background was black instead of white because I had my Mac set in Dark Mode. When I change it to Light Mode the background is white.

Great! Then it is not a bug but something I do not take into account when LenMus was designed, many years ago when dark modes and other color schemes were not common. LenMus was designed to use white background. I never thought about using dark mode or other user specific color schemes and I do not know the implications of allowing them. So I thing it would be better not to deal with this now and add support for dark and other modes for a future version. Then:
 
- For now, we need to investigate how to force LenMus to have white background.

- As a task for future, we need to identify all the changes to do for allowing dark and other color modes. If system colors are used, not only background color should change but also other colors (e.g.: text color, links colors, ...). Also, all images should have transparent background.

>  Sound works. I haven't given the program a thorough workout yet, but I will.

Perfect! If the test works there will be no problems with sound in other places.

> I have some difficulty resizing and iconizing windows. This must be a wxWidgets thing. I'll see if I can find more information on this.

If you need help send me a screenshot.

> I still get that annoying error window when I try to open a book

Yes, that's nothing to worry. The error is in the eBooks and I will fix it. But you will see this error until a new version of the books is uploaded.

I will update CMakeLists.txt so that cmake will create the macOS bundle. Also, if I add a couple of additional lines, cmake will use the bundle to create the package, and no more work will be needed. But as I cannot test CMakeLists.txt in macOS you will have to do all tests and needed changes, if any. I will do it now and inform you when I upload the changes.

**Comment by cecilios** [comment]
I've updated the eBooks to remove the bad value '20.0f'. Do you need help to update your repo?

Also, I have modified CMakeLists.txt to create the bundle, but I have not tested the changes as they are for macOS. So from this point, you should do any required fix. When it works, I expect that the macOS package could be generated as follows:
```
cmake .....     #generate the makefile
make            #compile and build
make install    #this will create the bundle, and then ...
make package    #this will generate the package
```

I attach the patch to apply to your CMakeLists.txt. To update it:
```
git apply --whitespace=fix patch.txt
```
[patch.txt](https://github.com/lenmus/lenmus/files/4565121/patch.txt)

**Comment by blurbler27** [comment]
I've given LenMus a more thorough look and most things seem to be working.  I went through several of the music reading exercises.

The lessons say that there should be a box near the metronome symbol to change the beats per minute.  I see no such box.  The metronome works fine though, except that I can't change the tempo.

The tests are done using the previous compile.  The Mac bundle version still needs some work.  For one thing, a template Info.plist needs to be provided.  Some (if not all) of the resource files need to be in the bundle.  I'll try to work on figuring this out.  I was able to get the app to run, but it couldn't find anything (books or icons.)

In one or two of my earlier tests, I did have the program freeze, so that a force-kill was necessary.  If I can get this going inside Xcode I can perhaps use debugging tools to figure out where this is happening.  (I don't recall the links right now, but will try to find them.)

I'm going to have to improve my Xcode skills to do this.  Most of my previous development experience has been in Linux and SunOS.

**Comment by cecilios** [comment]
> The lessons say that there should be a box near the metronome symbol to change the beats per minute. I see no such box. The metronome works fine though, except that I can't change the tempo.

There should appear two controls, as in this picture:
![metronome](https://user-images.githubusercontent.com/5238679/80991968-20a22a00-8e39-11ea-908e-74c8c25c67d1.png)

Clicking on the metronome icon should start/stop metronome sound. The box with the number is for setting metronme bpm. And the big T is for selecting which note type corresponds to a metronome beat. T means "as implied by time signature". Any other is self-explicative.

You have also a dedicated dialog for setting the metronome. Click on Menu > Tools > Metronome.

> a template Info.plist needs to be provided. Some (if not all) of the resource files need to be in the bundle. I'll try to work on figuring this out.

This should be now generated by the modifications to CMakeLists.txt that I sent in previous patch. I forget to attach Info.plist.in that I attach now. Place in in lenmus root folder. cmake should now generate the bundle and also the installation package. Perhaps you will have to fix something in CMakeLists.txt.

> If I can get this going inside Xcode I can perhaps use debugging tools

That would be great for debugging and programming!

Let me know if I can help in any way.

**Comment by blurbler27** [comment]
The area near the metronome icon looks like this:
![g](https://user-images.githubusercontent.com/32274358/80995657-38ab8680-8e04-11ea-9d58-3a917a22cd0a.png)

I tried accessing the metronome through the Menu > Tools > Metronome, and got this:
![mt](https://user-images.githubusercontent.com/32274358/80995749-61338080-8e04-11ea-89e8-f29fb5b78ae0.png)

HOWEVER, when I clicked the "continue" button, it went ahead and gave me the metronome controls:
![ms](https://user-images.githubusercontent.com/32274358/80995829-81633f80-8e04-11ea-8e51-b70c7d1f914f.png)

which I tried out, and everything seemed to work correctly.

Did you remember to attach your Info.plist.in?
Thanks,
Bob

**Comment by cecilios** [comment]
> Did you remember to attach your Info.plist.in?

Ahgg! forget again!. Here it is:
[Info.plist.txt](https://github.com/lenmus/lenmus/files/4576391/Info.plist.txt)

I had to change the extension to .txt  so that I can upload it here. Save it in lenmus root folder and change extension to .in.

I will review the trace and tell you something tomorrow. Now it is 8 pm here!

**Comment by cecilios** [comment]
To investigate why the toolbar does not displays the metronome controls you will have to review the code for creating these controls. The toolbars are created in `MainFrame::create_toolbars()` method (file src/app/lenmus_main_frame.cpp, line 1639). In line 1785 starts the code for creating the metronome controls.

The proble reported when trying to open the metronome tool dialog (Menu > Tools > metronome) is, apparently, that the index for setting the item to display in a `ComboBox` has an invalid value (out of range of defined values for that `ComboBox`). The warning is raised in method `DlgMetronome::create_dialog()` (file src/dialogs/lenmus_dlg_metronome.cpp, line 146). after reviewing the code I can not see anything wong. There are two `ComboBoxes` derived clases. In line 187 an instance of `wxChoice` is created and first item (item 0) is selected (line 188).
https://github.com/lenmus/lenmus/blob/fa62d8c45f90e694e9b9bdff21f270f55a30f62a/src/dialogs/lenmus_dlg_metronome.cpp#L187-L188
In line 323 an instance of `wxBitmapComboBox` is created and first item (item 0) is selected (line 326).
https://github.com/lenmus/lenmus/blob/fa62d8c45f90e694e9b9bdff21f270f55a30f62a/src/dialogs/lenmus_dlg_metronome.cpp#L323-L326
Values for these `ComboBoxes` are defined later, in methods `DlgMetronome::load_italian_tempi()` (line 353) and  `DlgMetronome::load_beat_notes()` (line 369)

Selecting item 0 after creating the combos is not a problem in Linux and Windows but, perhaps in wxWindows version for Mac, it is invalid to select item 0 before loading the combos. This could be the cause of the error. Try to comment out lines 188 and 326 as they are really not needed. I've have checked that removing them does not affect to other platforms.

**Comment by blurbler27** [comment]
Hi.  Sorry I've been busy with other things the last few days. 

Yes, commenting out the lines 188 and 326 from lenmus_dlg_metronome.cpp solved the problem occurring when trying to open the metronome tool dialog.

I've also succeeded in generating a lenmus.xcodeproj and getting the project opened in Xcode. Unfortunately, since the target is being placed in a location Xcode isn't expecting, I can't actually run the program from Xcode and use the debugging tools.  I'm still learning about Xcode and maybe I can figure this out.

I haven't really looked at why the toolbar doesn't contain all the metronome controls, but will try to get to it soon.

**Comment by cecilios** [comment]
Hi Bob,

> Sorry I've been busy with other things the last few days.

You do not need to apologize. We all have other things to do and this macOS task does not have any deadline!

> commenting out the lines 188 and 326 from lenmus_dlg_metronome.cpp solved the problem

Perfect! One less problem!

Being able to compile with Xcode is necessary for debugging and finding the origin of the problems. The task of preparing the environment to collaborate in programming and debugging is sometimes hard and maddening because we are forced to deal with tools we never used before and to deal with many new things. But don't worry too much. Little by little the problems are solved and in the end you get a lot of experience.

Did you have time to test the new cmake file? Does it generate the expected folders tree and the package?

Kind regards

**Comment by blurbler27** [comment]
The new cmake file still has a few bugs.  On running "make" the first time from the z_mybuild folder, it created the lenmus.app folder under z_mybuild/bin.  It correctly installed the Info.plist file under Contents and the executable under Contents/MacOS.

"make install" attempts to create a lenmus.app under /usr/local, so I had to sudo the make install.  It appears that the necessary libraries are in the Frameworks folder, but I will have to check that later.  (I'm going to try to install a fresh copy of Mac OS X under VirtualBox so that I can see if the app will work on a machine that doesn't have the extra libraries installed.)

However, the "make install" didn't include the share folder with the lenmus configuration files.  After I manually copied "share/lenmus" and its subfolders into the Contents folder, things appeared to work correctly.  I could see the books and exercises.

Right now, the remaining problems are 1) can't adjust the metronome settings from the toolbar, and 2) there are no buttons on the top level window for iconizing, maximizing, or closing the window.  This is probably a wxWidgets thing.

**Comment by blurbler27** [comment]
Two steps forward, one step back! 
I've tried to dust off my rusty git skills and work in a local branch (which I call macOS).  I can't upload this to the server so I am attaching the differences -- both the files and a git-diff file.  I'll try to figure out as much as possible myself, but if you have time to look, maybe something will occur to you.  I've attached the different files in a zip archive.
[macOSdiff.zip](https://github.com/lenmus/lenmus/files/4683221/macOSdiff.zip)
[git-diff.txt](https://github.com/lenmus/lenmus/files/4683235/git-diff.txt)

This includes the patched CMakeLists.txt file that creates a lenmus.app bundle, as well as some other changes I had to make to get it to compile.

This last attempt is made based on version 5.6.2.  (I downloaded a fresh version from the archive.) I was actually having better luck with the previous version.  I don't know if the changes you made caused this or if there is some other factor.

Current issues:

1.  When constructing the lenmus.app bundle, the bundle is expecting folders "Contents/share/lenmus/<version-number>/{locale,res,templates,xrc}" (I hope my abbreviation is understandable).  Neither "make" nor "make install" created these, so I put them in manually.
2.   lenmus.app runs after I do this, but the opening screen is no longer showing up.  I can see the menu for opening the books from the mac command tab, and it sees that the books are there, but they are not opening successfully.  (They were before.)
3.  I can open the project in Xcode by doing a "cmake <options> -G Xcode ../" from my build directory, and then opening the created "lenmus.xcodeproj" from Xcode.  It would be nice to be able to use Xcode's debugging facilities, but when I try to run lenmus from inside Xcode, it can't seem to find its configuration files.  I'm going to see if I can find a solution to this problem.

**Comment by blurbler27** [comment]
Sorry, I had forgotten the difference between "branch" and "fork".  I'll try to work in my own fork.  Will you be able to see this?

**Comment by cecilios** [comment]
You have a fork in your account at https://github.com/blurbler27/lenmus . A fork it is just a copy of the repo linked to it so that, if required, you can update your repo with changes in the main repo, and viceversa. Your fork is public and I can see it. Currently, your fork only has one branch, the main branch. But you can have as many you require.

It is perfectly OK to work on your fork and that's the expected way of doing things. When you are ready to send changes you just ask for a Pull Request on the LenMus main repo.

So, don't worry. Everything seems to be ok.

**Comment by cecilios** [comment]
Sorry! I didn't see this comment https://github.com/lenmus/lenmus/issues/103#issuecomment-634126328

I wil study it and reply later or tomorrow. Here it is 8.00 PM

**Comment by cecilios** [comment]
>  a local branch (which I call macOS). I can't upload this to the server

You cannot upload it to your remote repo because you have not created a remote branch, only a local branch. So do this:

1. Move to your local branch named 'macOS':
```
cd <path/to/your/projects/folder>/lenmus
git checkout macOS
```

2. Check that your local branch is linked to your remote fork:
```
git remote -v
git push -u origin macOS
```
It should say something as:
```
origin	git@github.com:blurbler27/lenmus.git (fetch)
origin	git@github.com:blurbler27/lenmus.git (push)
```

3. Push your changes from your local branch to your remote repo and force to create a remote branch named 'macOS' that will be syncronized with your local branch:
```
git push -u origin macOS
```

I will review now the files you send me and I will tell you something tomorrow.

**Comment by cecilios** [comment]
> When constructing the lenmus.app bundle, the bundle is expecting folders "Contents/share/lenmus//{locale,res,templates,xrc}" (I hope my abbreviation is understandable). Neither "make" nor "make install" created these, so I put them in manually.

`make` just builds the program. I does not install anything. Installation is done by `make install` (it does not build the program) it just install all files. Files to install and locations are specified in the CMakeLists.txt file. With reference to your CMakeLists.txt file in the 'maccOS' branch:

Here https://github.com/blurbler27/lenmus/blob/e89601a567bbce5bc3ff40e5c620295d23223edb/CMakeLists.txt#L381-L400 the installation folder ROOT_INSTALL is defined. It is set to default value in CMAKE_INSTALL_PREFIX. You can change that value when running cmake, with -DCMAKE_INSTALL_PREFIX=<the path you would like to install files>.

Therefore, after `make install` you should find everything in "user/share/lenmus//{locale,res,templates,xrc}". If you would like to find the files in "Contents/share/lenmus//{locale,res,templates,xrc}" you will have to run cmake like this:
```
cmake -DCMAKE_INSTALL_PREFIX="/Content/share/lenmus/" [...]
```
The files to copy and the subfolders are defined in differente places. Here https://github.com/blurbler27/lenmus/blob/e89601a567bbce5bc3ff40e5c620295d23223edb/CMakeLists.txt#L801-L809 the different subfolders for installation are defined.

And here https://github.com/blurbler27/lenmus/blob/e89601a567bbce5bc3ff40e5c620295d23223edb/CMakeLists.txt#L1247 start the installation, just cmake INSTALL commands to copy files to folders.

But `make install` is only needed when the user is building from sources. For preparing an installation package probably it would be better to use `make package`. This will triger these lines: https://github.com/blurbler27/lenmus/blob/e89601a567bbce5bc3ff40e5c620295d23223edb/CMakeLists.txt#L1538-L1540

I'm assuming that to generate the drang & drop package camke will use the same install paths that uses `make install` but I don't know

**Comment by cecilios** [comment]
> 2. Lenmus.app runs after I do this, but the opening screen is no longer showing up. I can see the menu for opening the books from the mac command tab, and it sees that the books are there, but they are not opening successfully. (They were before.)

I did not make any change in the source code related to the openning screen. The only difference in source code is to fix the path for the default soundfont. All other changes has nothing to do with source code but to CMakeLists.txt file and to some other files needed  to build AppImage packages and Debian packages.

**Comment by blurbler27** [comment]
Thanks for the comments, which are very helpful.  I'm still working on things, but here is a brief progress report.

1. cmake's DragNDrop is case sensitive, but was DRAGNDROP in the CMakeLists.txt file, so I had to edit that.
2. "make package" seems to try to be building the app.  Right now I'm stuck because cmake copies some of the necessary libraries into the Frameworks folder with permission read only (r--r--r--).  But then cmake tries to run the install_name_tool utility to fix up some location information, and of course it can't do that on read-only files.  Some of the libraries are copied with write permissions, though.  The ones I've spot checked are installed by homebrew packages, so I could probably change these permissions in their original install locations (using chmod) without consequence.  If this affects any of the Mac system libraries, I won't be able to do that.  (Apple is very protective of their installed software! I can't change things in the system folders even with sudo.)  I'll fix the ones I can and try again tomorrow, but no more time today.

I just found this link: [https://cmake.org/pipermail/cmake/2017-February/064962.html](url).  I'll look at this tomorrow.

**Comment by blurbler27** [comment]
The fix in the above link seems a little complex for now.  All the affected libraries are from homebrew installs, and apparently homebrew contributors aren't consistent in the permissions they put on their libraries.  For now, I'm just going to change the permissions of the installed library to read-write and go on.  
However, in the long run, you may wish to provide instructions for building LenMus on Mac OS, and others will run into this problem.  So this is just a short term fix.

**Comment by cecilios** [comment]
OK thank you. The important thing is to continue and get the package generated. Then we will have time to polish and fix details.

**Comment by blurbler27** [comment]
Another step forward and another step backwards.  Fixing the write permissions  allowed me to almost successfully create the package.  That is, it would create a disk image with a lenmus.app containing the lenmus executable under Contents/MacOS, and the needed libraries under Contents/Frameworks, and Contents/Info.plist.  However, none of the resources (books, images, etc.) were there.  I could run the app but it wouldn't find anything.

I modified the app by hand by adding all the resources under Contents/share/lenmus/5.6.2.  I then ran the app again, and in the File menu I saw the Open Books icon, and when I clicked on it, a window appeared listing the available books.  When I tried clicking on any of the books, nothing happened.  I won't say the program became unresponsive, because I could still exit it normally.

I feel that there's a path problem that I'm not understanding.  I've made some attempts to sift through your code to see how things work, but there's a lot of it!  Where in the code does LenMus figure out what folder contains its resources?  I'm also confused about the flags passed to cmake (-DCMAKE_INSTALL_PREFIX=...).  Does this refer to installation only, or does it tell LenMus where to look to find its files?

For the record, before trying to bundle LenMus into a MacOS app, things were working almost correctly, modulo a few minor problems, so I definitely think the path problem is the main thing.

**Comment by cecilios** [comment]
Do not despair that this is already well advanced and you are doing a great job. Thank you! The truth is that you are having a hard and frustrating job, because without knowing Xcode, wxWidgets, make and cmake, preparing a package for macOS becomes an odyssey. Most would have already abandoned.

> I'm also confused about the flags passed to cmake (-DCMAKE_INSTALL_PREFIX=...). Does this refer to installation only, or does it tell LenMus where to look to find its files?

Variable CMAKE_INSTALL_PREFIX is used in the cmake script to set up other variables:
https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L392-L393

`ROOT_INSTALL` is the used for copying files when you do `make install` and, I expect, when creating the Drang&Drop package.
`LENMUS_INSTALL_ROOT` is used for preparing the `lenmus_config.h` file that will be used during program compilation. Cmake does this as follows: https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L845-L849

In line 846 the command `configure_file` instructs cmake to open file lenmus_config.h.cmake, replace in it all variables betwwen '@' signs with the values set in the cmake script, and export the result as `lenmus_config.h`. As this file is part of lenmus sources, it will be used during lenmus compilation. In this way lenmus is informed about installation folders and other configuration options.

> Where in the code does LenMus figure out what folder contains its resources?

See file lenmus/src/globals/lenmus_paths.cpp. When the application is initialized one of the first thing it does is to create a `Paths` object. This object computes the paths to all resources. The code is well documented so I hope you will to not have problems to understand it.


> it would create a disk image with a lenmus.app containing the lenmus executable under Contents/MacOS, and the needed libraries under Contents/Frameworks, and Contents/Info.plist. However, none of the resources (books, images, etc.) were there. I could run the app but it wouldn't find anything.  ... I modified the app by hand by adding all the resources under Contents/share/lenmus/5.6.2

Uhm! this is strange. Resources are installed here:
https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L1412-L1434

And this should copy everything to ${RES_DIR}/*

eBooks are installed here: https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L1396-L1409

And book should be copied to ${LOCALE_DIR}/locale/*

In https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L789-L797

Uhm! I see you have changed CMakeList.txt from previous version! And you have removed the APPLE part for setting the paths. It was as follows:

```
# set installation paths
if(WIN32)
    #For CPack using NSIS, the DESTINATION path must be relative because installation path
    #is always specified at the moment when user executes the installer. Nevertheless
    #current NSIS installer does not allow the user to select instalation folder and
    #it is always installed in "C:/Program Files/lenmus"
    set(BIN_DIR bin)
    set(SAMPLES_DIR .)
    set(LOCALE_DIR .)
    set(LOCALE_LICENSES_DIR locale)
    set(RES_DIR res)
    set(TEMPLATES_DIR .)
    set(XRC_DIR .)
    set(DOCS_DIR .)
elseif(APPLE)
    set(BIN_DIR "${ROOT_INSTALL}/Contents/MacOS")
    set(SAMPLES_DIR $ENV{HOME}/${CMAKE_PROJECT_NAME} )
    set(LOCALE_DIR ${SHARED_DIR})
    set(LOCALE_LICENSES_DIR ${SHARED_DIR})
    set(RES_DIR "${SHARED_DIR}/res")
    set(TEMPLATES_DIR ${SHARED_DIR})
    set(XRC_DIR ${SHARED_DIR})
    set(DOCS_DIR ${SHARED_DIR})
else()
    set(BIN_DIR "${ROOT_INSTALL}/bin")
    set(SAMPLES_DIR $ENV{HOME}/${CMAKE_PROJECT_NAME} )
    set(LOCALE_DIR ${SHARED_DIR})
    set(LOCALE_LICENSES_DIR ${SHARED_DIR})
    set(RES_DIR "${SHARED_DIR}/res")
    set(TEMPLATES_DIR ${SHARED_DIR})
    set(XRC_DIR ${SHARED_DIR})
    set(DOCS_DIR ${SHARED_DIR})
endif()
```
With curren cmake code binaries are not isntalled in "${ROOT_INSTALL}/Contents/MacOS") but in "${ROOT_INSTALL}/bin".

RES_DIR is set to "${SHARED_DIR}/res" and LOCALE_DIR is set to ${SHARED_DIR}

And SHARED_DIR was set here:  https://github.com/blurbler27/lenmus/blob/ed803807d1f1adbdc04f3390590a566418047cad/CMakeLists.txt#L396-L403

Ahggg! Is is also changed!. It was like this:

```
#set shared folder
if(UNIX AND NOT APPLE)
    set(SHARED_DIR ${ROOT_INSTALL}/share/${CMAKE_PROJECT_NAME}/${LENMUS_PACKAGE_VERSION} )
elseif(APPLE)
    set(SHARED_DIR ${ROOT_INSTALL}/Resources )
elseif(WIN32)
    set(SHARED_DIR ${ROOT_INSTALL} )
endif()
```

With your current CMakeLists.txt file resources are not installed in "Content/Resources" but in other place, that could explain why lenmus does not finds them.

**Comment by blurbler27** [comment]
Thank you!  This clarifies a lot.  I may not have time to try to fix things today but will as soon as possible.

I did not explicitly remove the path information from CMakeLists.txt.  I think that what happened is that your original patch file was supposed to apply to a slightly modified version of CMakeLists.txt that I had created before.  But I started over with a freshly downloaded copy of the archive, and therefore a slightly different CMakeLists.txt file.  I decided that it would be better in the long run to make sure that any changes I made would not affect building on other platforms (e.g., by making substantial changes in if(APPLE) blocks.)  Consequently some of the hunks of the patch failed.  I thought that I had caught them and restored by hand, but missed that one. I should probably check the others.

**Comment by blurbler27** [comment]
I did have time to try something.  First, the version of CMakeLists.txt I was actually using was not the one in my repository!  It appears that I forgot to push my changes.  The one I was using actually had the path information set correctly.  I did a git push, so the macOS branch should be up to date now.

In my build directory, I did a "cmake -DMAN_INSTALL_DIR=~/devel/xcodev/   ../".  I found that I had to specify a MAN_INSTALL_DIR, otherwise installation would try to install the man page in a place it didn't have write permissions for.  (On a Mac, a user can't write in /usr/share even using the sudo command.)

Things are still not working correctly, but I don't have more time to work on it today.  I will get back to it soon.

**Comment by blurbler27** [comment]
I put some debugging statements in lenmus_paths.cpp to try to figure out where the executable is looking for its data.  Here's an example if install the app on my desktop:

oSharedHome /Users/rbyerly/Desktop/lenmus.app/Contents/share/lenmus/5.6.2
oLogsHome /Users/rbyerly/lenmus/temp
oConfigHome /Users/rbyerly/.config/lenmus/5.6.2
oDataHome /Users/rbyerly/lenmus/scores
oSoundFonts /Users/rbyerly/devel/xcodev/lenmus-3/lenmus/res/sounds

"make package" is **not** automatically copying share/lenmus/5.6.2 into the app bundle, but if I copy it manually at least it can find the listing of the books.  It still can't open them though.

oSoundFonts looks funny.  It's in the source tree.

After doing this debugging, I finally noticed where LenMus is putting its own log files (in ~/lenmus/temp) and am including a copy.  Do paths look right to you?
[rbyerly-lenmus-log.txt](https://github.com/lenmus/lenmus/files/4707712/rbyerly-lenmus-log.txt)

Incidentally, how are you including your code listings in your replies on this site?  Looks nice, but I don't see how you are doing it.

**Comment by cecilios** [comment]
> Incidentally, how are you including your code listings in your replies on this site? Looks nice, but I don't see how you are doing it.

References to some lines in a source file in the repo: https://help.github.com/en/github/managing-your-work-on-github/creating-a-permanent-link-to-a-code-snippet

Some lines of code, in a block: Beging and end the block with three backtics. See https://help.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax

> ... log files (in ~/lenmus/temp) and am including a copy. Do paths look right to you?

```
08:28:33: Binary root = /Users/rbyerly/Desktop/lenmus.app/Contents/MacOS
08:28:33: Shared root = /Users/rbyerly/Desktop/lenmus.app/Contents/share/lenmus/5.6.2/res/fonts/
08:28:33: Config root = /Users/rbyerly/.config/lenmus/5.6.2/
08:28:33: Data root = /Users/rbyerly/lenmus/scores/scores/
08:28:33: Logs root = /Users/rbyerly/lenmus/temp/
08:28:33: Sound fonts = /Users/rbyerly/devel/xcodev/lenmus-3/lenmus/res/sounds/
08:28:33: Source root = /Users/rbyerly/devel/xcodev/lenmus-3/lenmus/
```

For running LenMus any combination of paths is valid as long as the expected content is in those folders. But for creating the package (and also for running when installed from the package) the folders structure must be the structure required by macOS packages that, to my understanding (probably wrong) should be:
```
/Applications
    └── lenmus.app/
        └── Contents/
            ├── Info.plist      (generated by cmake. A template included)
            ├── version.plist   (should be generated by cmake)
            ├── PkgInfo (it only contains "APP???"  (echo "APPL????" > lenmus.app/Contents/PkgInfo)
            ├── Frameworks/
            │   └── <libname>.dylib (e.g.: wxWidges libraries)
            ├── MacOS/
            │   └── lenmus (the executable)
            └── Resources/      (copied here by “make install”)
                ├── locale/
                ├── res/
                ├── templates/
                ├── xrc/
                ├── AUTHORS
                ├── CHANGELOG.md
                ├── .. etc.
                │
```
So, paths reported in the log:

- Binary root `= /Users/rbyerly/Desktop/lenmus.app/Contents/MacOS` looks OK.
- Shared root `= /Users/rbyerly/Desktop/lenmus.app/Contents/share/lenmus/5.6.2/res/fonts/` looks wrong. It should be `= /Users/rbyerly/Desktop/lenmus.app/Contents/Resources/res/fonts/`
- Config root `= /Users/rbyerly/.config/lenmus/5.6.2/` is OK. Config file is not part of the package and is created in user home folder.
- Data root `= /Users/rbyerly/lenmus/scores/scores/` is OK. Samples are not part of the package and are copied to user home folder.
- Logs root `= /Users/rbyerly/lenmus/temp/` is OK. The folder for temporal files is not part of the package and is created in user home folder.
- Sound fonts `= /Users/rbyerly/devel/xcodev/lenmus-3/lenmus/res/sounds/` is wrong. It is using file from source tree and should be a file copied into the package structure, that is, should be `= /Users/rbyerly/Desktop/lenmus.app/Contents/Resources/res/sounds/`
- Source root `= /Users/rbyerly/devel/xcodev/lenmus-3/lenmus/. It is not important for the package. I assume it is correct. It is the pack to lenmus sources.

> "make package" is not automatically copying share/lenmus/5.6.2 into the app bundle ... oSoundFonts looks funny. It's in the source tree.

Both problems should be fixed in CMakeList.txt, perhaps when setting the paths, or when installing the files, or both!

Assuming your current CMakelIsts.txt file is the one in your repo, in branch 'macOS':

- The soundfont destination `DESTINATION ${LENMUS_ROOT_DIR}/res/sounds)` is bad. See https://github.com/blurbler27/lenmus/blob/710293e1109d0f4f7170b41810621e869eac3279/CMakeLists.txt#L848-L855
It should be `${SHARED_DIR}/res/sounds)`. Path for Bravura font has the same problem.

- The path for books looks OK. See https://github.com/blurbler27/lenmus/blob/710293e1109d0f4f7170b41810621e869eac3279/CMakeLists.txt#L1429-L1442

**Comment by blurbler27** [comment]
Just checking in.  I made the correction in CMakeFileList.txt that you suggested, but there are still problems in finding the files.
For now, I've reverted to playing with a command-line version of lenmus rather than the macOS bundle, and I finally learned the trick that lets me run it in the Xcode debugger.  Apple documents things fairly well for people who develop in Apple's preferred tools of Swift or Objective-C, but not for other systems,  particularly for make/cmake-based projects.  Fortunately, there are a lot of posts by programmers who have figured out some of these other issues.  Lenmus was being loaded by default as a "legacy build", and I only had to change the build type.
I have only limited time lately to work on this, but instead of playing around with various configuration options, I think I should now concentrate on fully understanding the setup logic. Being able to use the debugger to examine values of variables should help.

**Comment by cecilios** [comment]
Thanks for the feedback!

**Comment by blurbler27** [comment]
After playing around with the Xcode debugging facilities and examining the logical structure of the code, I think I see the main changes in lenmus_path.cpp that need to be made to get LenMus working in a MacOS bundle.  Right now the directory structure LenMus is expecting is incompatible with the bundle structure.
I notice in CMakeLists.txt that you set LENMUS_PLATFORM_MAC to 0 and LENMUS_PLATFORM_UNIX to 1 when building on an Apple platform.  The comment says that this can be changed after other issues are fixed.  Do you have a feeling for what issues might arise if we fix this?  On a Mac, it would be reasonable to use
```
set( LENMUS_PLATFORM_WIN32 "0")
    set( LENMUS_PLATFORM_UNIX "0")  
    set( LENMUS_PLATFORM_MAC "1")
```
or even 
```
set( LENMUS_PLATFORM_WIN32 "0")
    set( LENMUS_PLATFORM_UNIX "1") 
    set( LENMUS_PLATFORM_MAC "1")
```
if these won't cause trouble elsewhere.  I kind of like the second since macOS is basically a form of Unix.
Anyway, I am going to try to make the necessary changes in lenmus_path.cpp using preprocessor directives to avoid interfering with compilation on other platforms.

**Comment by cecilios** [comment]
LOMSE_PLATFORM_APPLE didn't existed in the lomse_config.h file. In march 15th, I introduced it just in case it was necessary and reviewed the usage of LENMUS_PLATFORM_UNIX so that all the code common for UNIX and APPLE were correctly identified. I didn't appreciate any need for LENMUS_PLATFORM_APPLE but left it there just in case it was necessary when you tested the app. and developed the package. LOMSE_PLATFORM_APPLE = "1" implies all other = "0". Only one of them should be "1".

When I modified CMakeList.txt set LOMSE_PLATFORM_APPLE to "0" as I have not identified any need, to prevent errors just in case I missed anything, that could interfere with your tests.

Feel free to play with this variable if needed and chage the code in lomse_paths but, in my oppinion, probably any necessary change in the paths should be achievable by modifying the settings in CMakeLists.txt. But this could be a later refinement once you can get the package.

**Comment by blurbler27** [comment]
Do you remember the purpose of the preprocessor test #ifndef __WXMAC__ in lenmus_app.cpp around lines 993?  (The line number is wrong since I've made a few edits to this file.  But it should be near there.)
```
#ifndef __WXMAC__
    m_frame->Show(true);
#endif //ndef __WXMAC__
    SetTopWindow(m_frame);
```
I wasn't getting anything until I commented those out and let m_frame->Show(true) run.
```
//#ifndef __WXMAC__
    m_frame->Show(true);
//#endif //ndef __WXMAC__
    SetTopWindow(m_frame);
```
Before making this change, the program would run but I wasn't seeing anything.  Now I can see the main frame and open the books again.  I will do some more testing as I have time.
I've pushed the changes to the macOS branch in my repository.  I know I've made changes in lenmus_app.cpp and lenmus_path.cpp to take care of the paths problem when running in a macOS app bundle.  Also in lenmus_updater.cpp to avoid using a call not implemented in macOS.
The metronome problem still exists, but I think we're getting closer to having a usable mac app.

**Comment by cecilios** [comment]
That code was taken from a wxWidgets sample. But as discussed here
https://github.com/lenmus/lenmus/issues/103#issuecomment-621733738
it seems useless. So you did it right commenting out those lines.

**Comment by blurbler27** [comment]
Ah yes.  I had forgotten your earlier post, but was trying to figure out for myself how things work.  (And I think I've made progress on that.)  I'll also take a look at the metronome problem, although you said you may rewrite that section.  It should be enlightening for me to see how it works even if you do.
Before distributing an app it would be nice to test it on some other mac.  I'm trying to install another copy of macOS in a VirtualBox for testing purposes but am having some difficulties.  Unfortunately most of my computer-literate friends are Windows or Linux users.  I could do a rough test by temporarily moving some of the dynamic libraries LenMus depends on to make sure that it is linking with the ones in the app bundle.

**Comment by cecilios** [comment]
Hi Bob, 

The metronome tool bar is set up here
https://github.com/blurbler27/lenmus/blob/f8cb5f477913e277379ab5069415dfe0843dfd2f/src/app/lenmus_main_frame.cpp#L1785-L1805

I remember having had problems to display some of the controls in Windows and this was solved when I added more space for them. Its seems to be that, depending on the platform, some controls takes much more space. At least, that was the problem I found in Windows and, probably it is the same problem now.

Look at lines 1795 and 1801. Instead of the default size I had to code a fixed amount of 60 pixels. You could try to increase this amount. Play also with the control height (variable nHeight). Replace that variable by a fixed number and do some tests to see if increasing the height solves the problem.

As to testing the package, if the package can now be built just from cmake without having to use Xcode, I know some people owning a Mac that could help on this. They know nothing about computers but if I give them clear instructions for building the package from command line, they could help. So, I would suggest that we prepare instructions for building and packaging.

Regards,

**Comment by blurbler27** [comment]
Hi Cecilios,
I tried playing with the number of pixels in lines 1795 and 1801 without success.  I noticed that wxSpinCtrl had been changed in wxWidgets 5.1, so I tried downgrading to wxWidgets 5.0 using the HomeBrew version.  No change either.  In fact, I was disappointed to see some strange effects when compiling using the HomeBrew version of wxWidgets.  Here's the opening screen:
![Lwx5 0start](https://user-images.githubusercontent.com/32274358/84787925-26e90180-afb4-11ea-97eb-619c84828a6f.png)

Notice the squished graphic on the left.  Some of the controls are in different places.  I also notated a few controls whose purpose I don't understand.  The version with wxWidgets 5.1 looks better.  This is with the most recent version of wxWidgets 5.1:
![Lwx5 1start](https://user-images.githubusercontent.com/32274358/84788083-5566dc80-afb4-11ea-991b-afd7df849085.png)

It's a pity the HomeBrew version is less functional.  That means anyone wishing to compile LenMus themselves on a Mac must also compile wxWidgets, and I did have some problems with that.

Still a few other issues.  For example, in General Exercises -- Other Exercises -- Clef Reading, if I click on "new problem", the program always crashes.  (I'm running this under the Xcode debugger, and the program terminates in line 140 of lomse_staffobjs_table.h (an inline function) with the message Thread 1: EXC_BAD_ACCESS (code=1, address=0x4).  I find online some references on the EXC_BAD_ACCESS message, so maybe I can get more information.)  It is possible for me to set breakpoints and trace the program a line at a time, so if you have some suggestion about where the best place to set a break point is, I can do it and examine the values of variables.  I have a much better understanding now of how the program works, but I have not yet penetrated very far into the structure of lomse.

The other exercises I've tried seem to work as expected, but I have only been spot-checking.  I'll try to go through them more systematically.

**Comment by cecilios** [comment]
Hi Bob, 

The pictures looks better than the ones in Dark Mode. Did you changed the mode of did you fixed the background color in LenMus?

For reference, this is what I get in Linux (in Windows it is similar). Notice that when only the start page is displayed only a few buttons in the toolbars are not grayed (the contrary that in macOS).
![lenmus-start-page](https://user-images.githubusercontent.com/5238679/84805268-47d94300-b004-11ea-9972-01703590aa8e.png)


As to your question, this toolbar 
![toolbar](https://user-images.githubusercontent.com/5238679/84805363-6dfee300-b004-11ea-8d07-3f095c12c6ef.png)
 is for score playback. In current version, you can open scores and play them. Menu > File > Open :  e.g. lenmus/samples/greensleeves_v16.lms 

Let's try to solve the problems one by one. Otherwise this will be really difficult.

I would suggest to use the wxWidgets version that comes with mac (or the one that HomeBrew installs). Unless there are important reasons, it is safer to stay with the version provided by the standard package for the current OS version. It is also simpler for building. So I would suggest moving back to wxWidgets 3.0

I have not seen any relevant change between wxWidgets 3.0 and 3.1 in wxSpinCtrl. To what changes do you refer?

I would propose to follow these steps:

- First thing to solve is to generate the package by using cmake and verify that it can run in another computer. We need to prepare some building/installation guidelines so that other people could help in the tests. It doesn't matter if the program has bugs and the GUI has problems. The point is that if the program can be build by others, the they can help spotting issues in the build process and in testing, to verify that the behaviour in other machines is the same. They can help in testing from a user point of view while you can focus on technical issues. Please, let's finish first this task before trying to solve the other issues.

- Once this is done, let's proceed solving issues, one by one. I would suggest to start with the GUI. There are many issues there:

    - The title bar after the window title. Why is it the window title repeated? What window/frame is responsible for that useless space on top? 
    - Different height for each tool bar. This is caused by the heights of the different controls (spin box, combo box, etc.) inserted on each toolbar. If the height is different probably the control has been inserted. Why is it not visible? Is there a missing show() method that we must invoke?
    - Order of tool bars. This is not important. If you observe, the toolbars are just reversed. This is a thing that happens in Windows in relation to Linux. The behaviour in macOS should be one or the other, unless there has been a change between wxWidgets 3.0 and 3.1. This is very simple to fix, as the program has an #if  (line 1813 in lenmus_main_frame.cpp) for choosing the order.
    - some buttons grayed. This is normal as they are not operational when the function is not relevant for the displayed content. In the case of the help button, this is a bug, common to all platforms. Help was removed but I forget to hide the button.

- Finally, once the package can be built and installed, and the GUI issues fixed, let's fix bugs in the code, such as the one causing the crash in General Exercises > Other Exercises > Clef Reading. I don't know what code are you using so it is difficult for me to determine the crash point. But I guess that lomse_staffobjs_table (L140) refers to this line:
```
inline int num_entries() { return m_numEntries; }
```
Or other inline method. The crash is caused because the ColStaffObjs object is nullptr. To see why, the simpler way is to look at the stack of calls to determine in which method the object was created, and from there to investigate why it is nullptr.

Anyway, as said, I propose to postpone all this issues and to finish first the issue of generating the package from cmake and preparing some build instructions.

Regards,
Cecilio

**Comment by blurbler27** [comment]
I'm only getting limited time to work on this lately, but here is a preliminary version of instructions for building on macOS.
[LenMusInstall.txt](https://github.com/lenmus/lenmus/files/4805123/LenMusInstall.txt)

Instructions refer to my branch since they rely on a few modifications that haven't yet been merged into the main repository.  I feel that I need to include a file that explains that this is not the main repository however.  I haven't yet done this, but I will do whatever you like on this.

The instructions seem a bit complex for someone who hasn't had much experience building software from source.  There are several steps that should be automated that I haven't had time to figure out how to do yet.

**Comment by cecilios** [comment]
Hi Bob,
Thank you for the file, but let's do share your work in the right way. You have done a hard work, you did some code modifications and you' wrote instructions. This work is not yet finished but the work done until now should be properly uploaded to LenMus repo. To do this, it is necessary that you prepare a Pull Request (PR) with the material you would like to upload. From my point of view this material is, at least:
- The build instructions for macOS (please, rename the file as "macOS-build-instructions.txt" and place it in forder 'docs/building').
- Any modified file that is relevant for this build instructions. Probably CMakeInstall.txt. Any other modified file? Some code in lenmus?

Do you know how to prepare the PR? If not, please tell me and I will gide you. Thanks!

**Comment by cecilios** [comment]
Great work with the PR!
Now there are build instructions and we share a common code base. And others could test your instructions!

Now, let's start solving the different issues you've found. But first, remember that as the main repo has been updated, you should also update your fork and your local branches. To do it:

Let's assume your are in your local macOS branch:

```
$ git stash                     #save any changes that you've done and are not yet committed

$ git checkout master           #move to master branch and
$ git fetch upstream            #get the changes from upstream

$ git merge upstream/master     #update your local master branch
	
$ git push			#synce. your remote master

$ git checkout macOS            ##return to working branch and update it
$ git merge master

$ git stash pop                 #recover your uncommitted changes

```
You should always do this when the master repo is updated by others. If there are no updates it doesn't harm.

If the repos has been updated and some of the changes conflict with your local changes (this should not be the case after your PR) you might get warnings and you might need to resolve the conflicts if you changed a file that has been changed in the main repo.

### Next steps 

As the hardest part (setting the environment, installing dependencies, building and creating the package, and writing build instructions) is done, what is left is to fix some issues. So, let's close this too long 103 issue "Package for macOS" and let's open specific issues for each found problem, to work on each, one by one.

So I'm closing this and opening the following ones (please, feel free to open new ones whenever you find something new that need to be fixed):
- macOS: Duplicated title bar after the window title (#113)
- macOS: Tool bar problems (#114)
- macOS: order of toolbars (#115)
- macOS: Crash in Clef Reading exercise (#116)

I would suggest to start with the order of toolbars (#115) as to fix it should be very easy.

**Comment by cecilios** [comment]
Better, I will leave this open, to remeber that we have to improve this.

**Comment by cecilios** [comment]
Hi Bob,
I've been reviewing your build instructions and I would propose some changes.

In 3.1 you wrote:
> There will be lots of warnings from the make command, but hopefully no errors.

We should fix all these warnings and remove that sentence. Could you please send me the log?. To generate it execute make but re-direct its output to a file: `make > log.txt`

In steps 3.2 and 3.3 you mention that this should be automated in future. I think that this is already automated: it is the `make package` step. So I would suggest to remove all that manual testing instructions as, in my oppinion, this is not necessary. What's the benefit? Better install and test. Moreover, by doing this probably you have not tested if `make package` alone works OK to build the package, as some of the files that this step should copy will be already in place when doing this. So I would suggest to test and ensure that this sequence of commands builds the package without problems:
```
        cd ......./lenmus
        mkdir mybuild
        cd mybuild
        cmake ..
        make package
```
Best regards,
Cecilio

**Comment by blurbler27** [comment]
Hi Cecilio

Unfortunately, "make package" still doesn't automate the creation of the
Resources folder in lenmus.app.  The manual step is still needed.

I am attaching the make.log.  I didn't capture the output of "make package"
to a file, but it's short, so I'm just copying it below.

Bob

__________________________________________________________________________________________________
output of "make package":

make package

[  1%] *Setting LenMus version information ...*

-- Found Git: /usr/bin/git (found version "2.24.3 (Apple Git-128)")

--   version git    = 'Release_5.6.2-6-g8ec26e6c'

LenMus version info:

-- LenMus version (LENMUS_VERSION) = '5.6.2'

-- LenMus version string (LENMUS_PACKAGE_VERSION) = '5.6.2'

-- LenMus build (LENMUS_VERSION_BUILD) = '8ec26e6c'

-- LenMus version long (LENMUS_VERSION_LONG) = '5.6.2+8ec26e6c'

[  1%] Built target lenmus-version

[100%] Built target lenmus

Run CPack packaging tool...

CPack: Create package using DragNDrop

CPack: Install projects

CPack: - Run preinstall target for: lenmus

CPack: - Install project: lenmus []

CPack: Create package

CPack: - package:
/Users/rbyerly/devel/xcodev/lenmus-3/lenmus/z_mybuild2/lenmus-0.1.1-Darwin.dmg
generated.

__________________________________________________________________________________________________

**Comment by cecilios** [comment]
Ok, thank you. This is something to fix in CMakeList.txt. I'll review it and I wiil tell you.

**Comment by cecilios** [comment]
Files that go into the `/Resources` folder are copied in CMakeLists.txt as follows:
Here
https://github.com/lenmus/lenmus/blob/ce121c08a4d2420e0d1d79d600a8b1dd430b7673/CMakeLists.txt#L407-L416

`SHARED_DIR` is set to `/Resources`

Later, at this point:
https://github.com/lenmus/lenmus/blob/ce121c08a4d2420e0d1d79d600a8b1dd430b7673/CMakeLists.txt#L804-L813

The different destination folders for the content are defined. Most of them go to SHARED_DIR

A few lines after, cmake logs the paths:
https://github.com/lenmus/lenmus/blob/ce121c08a4d2420e0d1d79d600a8b1dd430b7673/CMakeLists.txt#L824-L834

Could you please check the log?

Everything seems to be correct. Perhaps the problem is that the destination folder must exit and, if not, cmake fails to copy the files. I will investigate this possibility. Perhaps there is also something about in the cmake log that you can spot.

**Comment by cecilios** [comment]
Hi Bob,
I think I have discovered the problem. The default option in CMakeList.txt script is to build just the program and not to copy all other required files. See CMakeLists.txt, line 67. To check that this is the problem, please run
```
cd ......../lenmus
mkdir mybuild (if the folder already exist, clear all content)
cd mybuild
cmake ..
```

In the log you should see some messages, probably with this content:
```
Packages to build:
BUILD_PKG_MAIN = ON
BUILD_PKG_COMMON = OFF
BUILD_PKG_I18N = OFF
BUILD_PKG_EBOOKS = OFF
BUILD_PKG_BOUNDLE = OFF
```

Package main should be ON and the others off, meaning that only the program is built.  Thus, for creating the boundle package the invokation of cmake should be as follows:
```
cmake -DBUILD_PKG_BOUNDLE=ON ..
```
Please check if this solves the problem. Thank you!

**Comment by blurbler27** [comment]
Hi Cecilio,
Unfortunately, we're not there yet.  I tried `cmake -DBUILD_PKG_BOUNDLE=ON ..` from a clean build directory, and then `make` and `make package`.

After the `cmake -DBUILD_PKG_BOUNDLE=ON ..`, I saw the output:
```
Packages to build:
-- BUILD_PKG_MAIN = OFF
-- BUILD_PKG_COMMON = OFF
-- BUILD_PKG_I18N = OFF
-- BUILD_PKG_EBOOKS = OFF
-- BUILD_PKG_BOUNDLE = 1
```
but the Resources folder was still not showing up in the package.

**Comment by cecilios** [comment]
Could you please send me the cmake log?

**Comment by blurbler27** [comment]
[cmake.log.txt](https://github.com/lenmus/lenmus/files/4832122/cmake.log.txt)
This was using ``cmake -DBUILD_PKG_BOUNDLE=ON ..``

**Comment by cecilios** [comment]
Thank you Bob.

Looking at the log:
```
-- ROOT_INSTALL = /usr/local
-- BIN_DIR = /usr/local/Contents/MacOS
-- SHARED_DIR = /usr/local/Resources
-- SAMPLES_DIR = /Users/rbyerly/lenmus
-- LOCALE_DIR = /usr/local/Resources
-- LOCALE_LICENSES_DIR = /usr/local/Resources
-- RES_DIR = /usr/local/Resources/res
-- TEMPLATES_DIR = /usr/local/Resources
-- XRC_DIR = /usr/local/Resources
-- DOCS_DIR = /usr/local/Resources
```
It seems that the `Resources` folder (variable SHARED_DIR) is created at the same level than the `Contents` folder, instead of being a subfolder of `Contents`. So the fix is to change, in CMakeList.txt, line 411:
https://github.com/lenmus/lenmus/blob/b295a419b10375539853154fb8a83dea752b7525/CMakeLists.txt#L410-L411

to
```
set(SHARED_DIR ${ROOT_INSTALL}/Contents/Resources )
```

Please test this. Thank you.

**Comment by blurbler27** [comment]
No, that's not it either.

The ``lenmus.app`` folder that is used by ``make package`` to create the DMG file is inside the build directory.  The variables ``ROOT_INSTALL``, ``SHARED_DIR``, etc.  are entirely ignored.  "make" is installing files in these places, but they're not being used to create the package.  But anything I put in ``z_mybuild/bin/lenmus.app`` by hand will wind up in the DMG package.

**Comment by cecilios** [comment]
Sorry Bob, I do not understand clearly what you say. You say that:

>  "make" is installing files in these places, ...

So, you start with an empty build folder. Then you run `cmake`. Now you run `make` and after this the `lenmus.app` folder, contains the right structure and all the needed files?

But
>  but they're not being used to create the package

Then you run `make package` but it does not include the files in the package unless you copy again these files manually?

**Comment by blurbler27** [comment]
What we see in the cmake log:
```
-- ROOT_INSTALL = /usr/local
-- BIN_DIR = /usr/local/Contents/MacOS
-- SHARED_DIR = /usr/local/Resources
-- SAMPLES_DIR = /Users/rbyerly/lenmus
-- LOCALE_DIR = /usr/local/Resources
-- LOCALE_LICENSES_DIR = /usr/local/Resources
-- RES_DIR = /usr/local/Resources/res
-- TEMPLATES_DIR = /usr/local/Resources
-- XRC_DIR = /usr/local/Resources
-- DOCS_DIR = /usr/local/Resources
```
is pretty much all wrong for a package install.  "make" first builds the lenmus executable in a subdirectory of the build directory (e.g., z_mybuild/bin/lenmus.app/Contents/MacOS).  The ``Info.plist`` is also placed in z_mybuild/bin/lenmus.app/Contents.  I've found by trial-and-error that anything that gets placed in z_mybuild/bin/lenmus.app  will go into the app bundle produced by "make package".  The stuff that goes into /usr/local is ignored.  (Even, though, oddly enough, "make package" will put some stuff in /usr/local.  AFAIK it's just not used.)

It's getting late here and I don't have time tonight, but I think I'll try ``cmake -DCMAKE_INSTALL_PREFIX=./bin/lenmus.app/Contents..`` next, and see what happens.

**Comment by cecilios** [comment]
Hi Bob, 
Let me explain briefly how cmake & make tool chain works:

- cmake just creates a makefile for the tool specfied in the generator parameter. As we do not specify any, cmake then generates the default makefile for the macOS platform. This is a makefile to be processed by 'make' tool. It is important to notice that cmake has to be run in an empty build folder; otherwise many of the obsolete options from previous run will be there and will not be updated by the new run. And you will not get the expected results.

- the makefile is just a list of rules that say, in words:
    - to compile do 'this'; to do 'this' there are no prerequisites
    - to install do 'that'; before doing 'that' it is necessary to do 'this'
    - to package do foo; before doing 'foo' it is necessary to do ...

- So when we do `make` it will do the default action, normally to compile and to build the program, but will not copy files to directories.
- `make install` normally just copies files to the specified destinations to install a program. And normally has as prerequisite to do `make` so it if not done, `make install` will do both: compile and build the program and then copy files.
- Finally, `make package` normally requires `make` and `make install` and does the action of bundling all. What to do depends. Sometimes it is to copy all installed files in a zip file; others, in a debian package; or in any other way specified in the makefile.

Now lets' analyse the data we have:
- You say, in previous post, that in the cmake long you see:
```
-- SHARED_DIR = /usr/local/Resources
```
But if you followed my instructions and you did change line 411 of CMakeLists.txt, that line should be:
```
-- SHARED_DIR = /usr/local/Contents/Resources
```
As I'm totally sure you did change CMakeLists.txt, the most probably cause for not displying the new path is that you run cmake without totally clearing the build folder and, thus, the old options are in place and were not updated.

Nevertheless, you tell me that:
> I've found by trial-and-error that anything that gets placed in z_mybuild/bin/lenmus.app will go into the app bundle produced by "make package"

Did is not what I expected. Normally, when preparing a package, cmake copies the files to the package from the paths to install. But I don't know how 'DragNDrop' cmake package works and, as you have experimentally discovered, it seems to copy files from the build directory.

So, what we should try now is to install the files not in '/usr/local/..' but in the build directory. Therefore, the test you propose `cmake -DCMAKE_INSTALL_PREFIX=./bin/lenmus.app/Contents` could work, but pay attention to the paths reported for SHARED_DIR. Perhaps it is necessary to fix the passed CMAKE_INSTALL_PREFIX.

**Comment by blurbler27** [comment]
Hi Cecilios,
I'm pretty familiar with how make works, although not with cmake.  And the fact that the makefiles produced from cmake will call further cmake scripts makes their behavior somewhat confusing!

I noticed some time ago that it is necessary to clear the build directory before re-running cmake, and I have been doing that (using "rm -r  *" from the build directory) so that's not the problem.

I don't know if I'll have time today to do more due to family obligations, but I tried a quick run setting CMAKE_INSTALL_PREFIX to ./bin/lenmus.app/Contents, and it did install the Bravura fonts (although with a path problem -- probably easy to fix) but nothing else.  

As an experiment, I also tried running "make install", thinking that it might install the support files in the build copy of the lenmus app.  The problem there is that "make install" also tries to create a copy of the lenmus.app folder in its target directory, so I was getting a sequence of folders: ``lenmus.app/Contents/lenmus.app/Contents/ ...`` until the path names got too long and the script stopped.

I have tried looking online for further information about how the cmake DragNDrop feature works, since I think that this is the key.  I'm going to read more of the cmake documentation.
